### ISSUE TITLE
Build failures on OpenBSD

### ISSUE THREAD

**Issue by cfallin** [bytecodealliance/rsix#32]
Unfortunately it looks like there have been some regressions since #11 (completely understandable as OpenBSD is not tested on CI and is quite an odd duck wrt platform APIs).

The build failures are mostly undefined symbols in `libc`; here is the full build log: [link](https://gist.github.com/cfallin/7ccade106d5ba97da24373d14e68c3b6)

(This was using Rust 1.54.0 on OpenBSD 7.0-current; the current stable, 6.9, has an old Rust 1.51.0 that fails to build a bunch of stuff, including `unsafe-io`.)

I'm not sure how much work it would take to resolve these and I can't put much time into it at the moment but I figured I should record these in an issue.

**Comment by sunfishcode** [comment]
Thanks for the report! I don't have a convenient way to test OpenBSD compilation, but I submitted #33 which hopefully fixes all the errors listed in that log.

**Comment by cfallin** [comment]
Thanks! #33 almost does it; I fixed a few more issues with this patch:

```
diff --git a/src/imp/libc/fs/dir.rs b/src/imp/libc/fs/dir.rs
index cde4231..81ef752 100644
--- a/src/imp/libc/fs/dir.rs
+++ b/src/imp/libc/fs/dir.rs
@@ -103,7 +103,6 @@ unsafe fn read_dirent(dirent_ptr: *const libc_dirent) -> libc_dirent {
         target_os = "ios",
         target_os = "macos",
         target_os = "netbsd",
-        target_os = "openbsd",
         target_os = "wasi",
     )))]
     let d_off = (*dirent_ptr).d_off;
diff --git a/src/imp/libc/io/types.rs b/src/imp/libc/io/types.rs
index 908d741..2d2b8a1 100644
--- a/src/imp/libc/io/types.rs
+++ b/src/imp/libc/io/types.rs
@@ -104,7 +104,7 @@ bitflags! {
         )))]
         const FIXED_NOREPLACE = libc::MAP_FIXED_NOREPLACE;
         /// `MAP_GROWSDOWN`
-        #[cfg(not(any(target_os = "netbsd", target_os = "ios", target_os = "macos", target_os = "freebsd", target_os = "redox")))]
+        #[cfg(not(any(target_os = "netbsd", target_os = "openbsd", target_os = "ios", target_os = "macos", target_os = "freebsd", target_os = "redox")))]
         const GROWSDOWN = libc::MAP_GROWSDOWN;
         /// `MAP_HUGETLB`
         #[cfg(not(any(
```

Unfortunately it seems there's an issue in `libc` in which `libc::dirent64` grew a private field so [this](https://github.com/bytecodealliance/rsix/blob/aaf1da52cf62440ef921bfcba19a4e9fd3f21559/src/imp/libc/fs/dir.rs#L136) literal that builds a `libc::dirent64` is simply not possible to write with the current version of `libc`:

```
error[E0451]: field `__d_padding` of struct `dirent` is private
   --> src/imp/libc/fs/dir.rs:173:9
    |
173 |         __d_padding: zeroed(),
    |         ^^^^^^^^^^^^^^^^^^^^^ private field

```

it seems this is defined [here](https://github.com/rust-lang/libc/blob/1ba11bb2ffb8e155ed1b305f7723c0c41edba78a/src/unix/bsd/netbsdlike/openbsd/mod.rs#L432) and has been that way for 3 years, so I think the direct construction of a `libc::dirent64` is what's new here. It looks like the bindings in `libc` generally don't expose padding fields so it might be necessary to go about this struct-size fixup another way, though nothing immediately comes to mind as a cleaner alternative.

**Comment by sunfishcode** [comment]
I've now fixed these in #34.

**Comment by sunfishcode** [comment]
These are now fixed in #34, and we now have `cargo check`-level testing in CI in #36, which will catch all of these kinds of regressions going forward.
